processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,481 @@
+        // because nextReadIndex is changed only by the reading thread.
+        return !finished || remaining() > 0;
+      }
+
+      /**
+       * Calculate the number of remaining items in the array.  Must be called only when
+       * {@link StreamingMergeSortedGrouper#finished} is true.
+       *
+       * @return the number of remaining items
+       */
+      private int remaining()
+      {
+        if (curWriteIndex >= nextReadIndex) {
+          return curWriteIndex - nextReadIndex;
+        } else {
+          return (maxNumSlots - nextReadIndex) + curWriteIndex;
+        }
+      }
+
+      @Override
+      public Entry<KeyType> next()
+      {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+
+        final int recordOffset = recordSize * nextReadIndex;
+        final KeyType key = keySerde.fromByteBuffer(buffer, recordOffset);
+
+        final Object[] values = new Object[aggregators.length];
+        for (int i = 0; i < aggregators.length; i++) {
+          values[i] = aggregators[i].get(buffer, recordOffset + aggregatorOffsets[i]);
+        }
+
+        final int increaseTo = nextReadIndex == maxNumSlots - 1 ? 0 : nextReadIndex + 1;
+        // Wait if the array is empty until at least one slot becomes available for read, and then increase
+        // nextReadIndex.
+        increaseReadIndexTo(increaseTo);
+
+        return new Entry<>(key, values);
+      }
+
+      /**
+       * Wait for {@link StreamingMergeSortedGrouper#curWriteIndex} to be moved if necessary and move
+       * {@link StreamingMergeSortedGrouper#nextReadIndex}.
+       *
+       * @param target the target index {@link StreamingMergeSortedGrouper#nextReadIndex} will move to
+       */
+      private void increaseReadIndexTo(int target)
+      {
+        final long startAt = System.nanoTime();
+        final long spinTimeoutAt = startAt + SPIN_FOR_TIMEOUT_THRESHOLD_NS;
+        long now = startAt;
+
+        // In the below, we check that the array is empty and wait for at least one slot to become available. This is
+        // done in a while loop instead of using a lock to avoid frequent thread park.
+        //
+        // curWriteIndex is a volatile variable and the changes on it should be continuously checked until they are seen.
+        // See the following links.
+        // * http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.3.1.4
+        // * http://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5
+        // * https://stackoverflow.com/questions/11761552/detailed-semantics-of-volatile-regarding-timeliness-of-visibility
+        while ((!isReady() || target == curWriteIndex) && !finished && !Thread.currentThread().isInterrupted()) {
+          if (now >= queryTimeoutAtNs) {
+            throw new RuntimeException(new TimeoutException());
+          }
+          // Thread.yield() should not be called from the very beginning
```

Comment:
Please insert `ThreadHints.onSpinWait()`, see the Chronicle Map's source mentioned above as example as well

Reply:
Thread.onSpinWait() is added in java 9.

Reply:
ThreadHints is a polyfill

Reply:
Yes, but is it necessary in this pr? It won't be called until we support java 9.

Reply:
I think it's better to add now, I can bet nobody will return to this code to add onSpinWait() in 6  mo/1 year.

Reply:
I'll add later.